### PR TITLE
Add test for the blog generator

### PR DESCRIPTION
**Pull request by PropheticBird**

### CODE REVIEW

File: tests/ext/generators/test_blog.py [hunk 1]
```diff
@@ -0,0 +1,317 @@
+# coding: utf-8
+"""
+    tests.ext.generators.test_blog
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+    Tests Blog generator.
+
+    :copyright: (c) 2014 by the Holocron Team, see AUTHORS for details.
```
Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: `(c) 2015`

File: tests/ext/generators/test_blog.py [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+
+    def test_tags_with_posts(self):
+        """
+        Test that tags function writes a post to a tag template.
+        """
+        content = self._get_content([self.post_early], self.blog.tags)
+
+        self.assertIn('<h2>2012</h2>', content)
+
+    # Test feed
```
Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: I would like to separate test cases for feed and others. I mean, you can move all test cases related to feed into some separate class (e.g. TestBlogFeed). I think, it will increase readability.

File: tests/ext/generators/test_blog.py [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+                        'save_as': 'myindex.html',
+                    },
+                },
+            },
+        })))
+
+        # load template that is used to list posts in index and tags generators
+        templates_path = os.path.join(
+            os.path.dirname(app.__file__), 'themes', 'default', 'templates')
+        env = Environment(loader=FileSystemLoader(templates_path))
```
Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: I don't like that you manually prepare a Jinja2 env. I'd like to not perform such actions. As far as I understand, you do this because the Blog generator requires it, right? I propose to pass Holocron instance instead of mock object.

File: tests/ext/generators/test_blog.py [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+        self.assertNotIn('<h2>', content)
+
+    def test_index_with_posts(self):
+        """
+        Index function has to form an html files, where all posts are displayed
+        and grouped by year, from the newest on top, to the oldest in the
+        bottom.
+        """
+        content = self._get_content(
+            # assume that extract_posts sorted the posts by date
```
Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: Looks like this comment lies. BTW, I'd like rewrite this two statements into 

```
        posts = [self.post_early, self.post_moderate, self.post_late]
        content = self._get_content(posts, self.blog.index)
```

or

```
        content = self._get_content(
            [self.post_early, self.post_moderate, self.post_late], 
            self.blog.index)
```

File: tests/ext/generators/test_blog.py [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+            [self.post_early, self.post_moderate, self.post_late],
+            self.blog.index
+        )
+
+        h_year = '<h2>{0}</h2>'.format
```
Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: Сцуко жжжошь :D

File: tests/ext/generators/test_blog.py [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+        Tags function has to save a tag html representation to a correct
+        location, which is fetched from the configuration file.
+        """
+        with mock.patch(self.open_fn, mock.mock_open(), create=True) as mopen:
+            self.blog.tags([self.post_early])
+
+            self.assertEqual(
+                mopen.mock_calls[0][1][0],
+                'path/to/output/mypath/tags/testtag1/myindex.html'
+            )
```
Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: Please move `)` to the line above.

File: tests/ext/generators/test_blog.py [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+            # assume that extract_posts sorted the posts by date
+            [self.post_early, self.post_moderate, self.post_late],
+            self.blog.index
+        )
+
+        h_year = '<h2>{0}</h2>'.format
+
+        self.assertIn(h_year(2014), content)
+        post_position = content.index(h_year(2014)) + len(h_year(2014))
+        content = content[post_position:]
```
Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: Why this slice?

Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: Oh.. that's a kind of checking posts order?

File: tests/ext/generators/test_blog.py [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+        Test that feed writes correct values to an xml template.
+        """
+        content = self._get_content([], self.blog.feed)
+        content = self._xml_to_dict(content)
+
+        self.assertEqual('MyTestSite Feed', content['feed'][0]['title'])
+        self.assertEqual('http://www.mytest.com/', content['feed'][2]['id'])
+        self.assertEqual('http://www.mytest.com/myfeed.xml',
+                         content['feed'][3]['link'])
+        self.assertEqual('http://www.mytest.com/', content['feed'][4]['link'])
```
Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: What are the indices? I mean `[0]`, `[2]`, `[3]` and `[4]` ?

File: tests/ext/generators/test_blog.py (hunk 3 above)
Comment (PropheticBird) [inline tests/ext/generators/test_blog.py]: Why creating a test Jinja2 env sounds like a bad idea for you? 
Yes, you were right, it is because of the Blog generator. I thought, we are testing Blog generator code, but not the holocron.app code, so testing env will fit in here.

File: tests/ext/generators/test_blog.py (hunk 7 above)
Comment (PropheticBird) [inline tests/ext/generators/test_blog.py]: Yeah, we expect the newer post to appear first, then cut the string an perform another assertion.
It's dirty, I know.

File: tests/ext/generators/test_blog.py (hunk 4 above)
Comment (PropheticBird) [inline tests/ext/generators/test_blog.py]: OK, comment fixed.

Btw, it seems like posts are sorted not only by the `extract_posts`, but in the template as well.

File: tests/ext/generators/test_blog.py (hunk 6 above)
Comment (PropheticBird) [inline tests/ext/generators/test_blog.py]: OK.

File: tests/ext/generators/test_blog.py (hunk 8 above)
Comment (PropheticBird) [inline tests/ext/generators/test_blog.py]: These are the indices of the list, which contains the `title`, `link` etc dictionaries.
I made a mistake in the code that builds dict from xml. I'll fix it.

File: tests/ext/generators/test_blog.py (hunk 1 above)
Comment (PropheticBird) [inline tests/ext/generators/test_blog.py]: It's already `2015`.. *sigh*

File: tests/ext/generators/test_blog.py (hunk 3 above)
Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: Because if you somehow change this context in app.py, you will also need a change (**the same change**) here. Moreover, there's a ticket in issue tracker to move on abstract template engine. So I'd prefer to do not hardcode this.

File: tests/ext/generators/test_blog.py (hunk 4 above)
Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: Sure they are. Do you forget a hack with grouper in template?

File: tests/ext/generators/test_blog.py [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+                    if is_attribute(attr):
+                        url_data[attr.nodeName] = attr.getAttribute('href')
+                    else:
+                        url_data[attr.nodeName] = attr.firstChild.nodeValue
+
+                posts.append(url_data)
+                entries[url.nodeName] = posts
+
+        content = {root.nodeName: entries}
+
```
Comment (PropheticBird) [inline tests/ext/generators/test_blog.py]: This code is ugly, but currently it's the best I can produce. I tried to use dictionary comprehensions with `if-else` clauses, but it didn't work out. 

Can you make any suggestions about how to improve this?

**Comment by PropheticBird** [comment]
I've put each "sub-generator" to its corresponding class and provided a base class to set up common data for test. Not sure this is the correct way to do this. 
Maybe it is possible to create this test data once and use it after, since we only use it, but don't change, what do you think?

File: tests/ext/generators/test_blog.py [hunk 10]
```diff
@@ -0,0 +1,348 @@
...
+from dooku.conf import Conf
+from holocron.ext.generators import blog
+from holocron.content import Page, Post, Static
+from holocron.app import Holocron
+
+from tests import HolocronTestCase
+
+
+class TestBlogGenerator(HolocronTestCase):
```
Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: I'd like to rename this class. You know, by default almost all test runners discover all beings started with *test* as tests, but this is not a test. It's a helper, actually. So I'd like to call it `BlogTestCase` or something like that.

File: tests/ext/generators/test_blog.py [hunk 11]
```diff
@@ -0,0 +1,348 @@
...
+    some helper methods are provided.
+    """
+
+    def setUp(self):
+        """
+        Creates generator object and number of posts which will be further used
+        in test sets.
+        """
+
+        self.blog = blog.Blog(Holocron(conf=Conf({
```
Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: I forgot to discuss with you a [new format for declaring configuration data](https://github.com/ikalnitsky/holocron/blob/cbddea431d68e97826351761323c26b12b0c2884/holocron/app.py#L78-L119).

Actually, I didn't use it before, but it's a less verbose and navigation a bit easier now. Still, I'm not sure. What do you think?

File: tests/ext/generators/test_blog.py [hunk 12]
```diff
@@ -0,0 +1,348 @@
...
+            return content
+
+
+class TestExtractPosts(TestBlogGenerator):
+    """
+    Test extract posts function.
+    """
+
+    def setUp(self):
+        TestBlogGenerator.setUp(self)
```
Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: What a hell is it?) Please, don't repeat this common mistake again.

File: tests/ext/generators/test_blog.py [hunk 13]
```diff
@@ -0,0 +1,348 @@
...
+        Test extract_posts should process correctly empty lists.
+        """
+        documents = self.blog.extract_posts([])
+
+        self.assertEqual(len(documents), 0)
+
+    def test_extract_posts_with_all_doc_types(self):
+        """
+        The extract_posts has to separate posts from all other types
+        of dicuments passed as input.
```
Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: typo: d`i`cuments.

File: tests/ext/generators/test_blog.py [hunk 14]
```diff
@@ -0,0 +1,348 @@
...
+        self.assertEqual(documents[1], self.post_moderate)
+        self.assertEqual(documents[2], self.post_early)
+
+
+class TestIndexGenerator(TestBlogGenerator):
+    """
+    Test index generator.
+    """
+
+    def setUp(self):
```
Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: Meaningless declaration.

File: tests/ext/generators/test_blog.py [hunk 15]
```diff
@@ -0,0 +1,348 @@
...
+
+        self.assertIn(h_year(2012), content)
+
+
+class TestTagGenerator(TestBlogGenerator):
+    """
+    Test tag generator.
+    """
+
+    def setUp(self):
```
Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: Same as for previous test class.

File: tests/ext/generators/test_blog.py [hunk 16]
```diff
@@ -0,0 +1,348 @@
...
+
+        self.assertIn('<h2>2012</h2>', content)
+
+
+class TestFeedGenerator(TestBlogGenerator):
+    """
+    Test feed generator.
+    """
+
+    def setUp(self):
```
Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: Same as for previous test class.

File: tests/ext/generators/test_blog.py [hunk 17]
```diff
@@ -0,0 +1,348 @@
...
+
+    def test_save_index_to_filesystem(self):
+        """
+        Test that index function saves index.html file to a proper location.
+        """
+        with mock.patch(self.open_fn, mock.mock_open(), create=True) as mopen:
+            self.blog.index([])
+
+            self.assertEqual(
+                mopen.call_args[0][0], 'path/to/output/myindex.html')
```
Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: Actually, index generator shouldn't support filename changes. It should be exactly an `index.html` in the output directory; otherwise - it makes no sense to generate this. 

I'm not sure if it's related to this patchset, but I think we definitely should remove this. I mean both testcase and business code. I could be done in separate patch. If you decide so - please file a ticket in issue tracker.

File: tests/ext/generators/test_blog.py [hunk 18]
```diff
@@ -0,0 +1,348 @@
...
+
+    def test_template_variables(self):
+        """
+        Test that index function populates template with corrent data, fetched
+        from the configuration settings.
+        """
+        content = self._get_content([], self.blog.index)
+
+        self.assertIn('<title>MyTestSite</title>', content)
```
Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: Are there only two variables?

File: tests/ext/generators/test_blog.py [hunk 19]
```diff
@@ -0,0 +1,348 @@
...
+        self.assertIn('<p>&copy; 2014, Tester</p>', content)
+
+    def test_index_empty(self):
+        """
+        Test that running input on empty list of documents doesn't cause crash.
+        """
+        content = self._get_content([], self.blog.index)
+
+        self.assertNotIn('<h2>', content)
```
Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: Please, add comment that `<h2>` is an HTML tag for post titles. It'd help for other code readers.

Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: Or, damn.. `<h2>` is not about post titles. It's about years, right? We definitely need a comment. :)

P.S: perhaps even we can move it to test's class variable:

```python
class TestIndexGenerator(...):
    year_tag = '<h2>'
    posttitle_tag = 'bla-bla
````

File: tests/ext/generators/test_blog.py [hunk 20]
```diff
@@ -0,0 +1,348 @@
...
+
+    def test_index_with_posts(self):
```
Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: I'd like to see here and in the testcase above the checks for post titles.

File: tests/ext/generators/test_blog.py [hunk 21]
```diff
@@ -0,0 +1,348 @@
...
+        content = content[post_position:]
+
+        self.assertIn(h_year(2013), content)
+        post_position = content.index(h_year(2013)) + len(h_year(2013))
+        content = content[post_position:]
+
+        self.assertIn(h_year(2012), content)
+
+
+class TestTagGenerator(TestBlogGenerator):
```
Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: Actually, this testsuite didn't cover corner cases. It'd be cool to see:

* check that we write into two different files if a post have two tags
* check that we write into two different files if we have two posts with different tags
* check that we write into one file if we have  two posts with same tag
* check that written files contain urls for posts

File: tests/ext/generators/test_blog.py (hunk 20 above)
Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: Or even better - for post urls. I think it's more important to check urls.

File: tests/ext/generators/test_blog.py (hunk 11 above)
Comment (PropheticBird) [inline tests/ext/generators/test_blog.py]: I didn't see your pull request while writing this code. 

Well, I personally didn't like this format at a first glance, I think it decreases readability. But maybe this is my habit speaking, so we can give it a try and change it later if we don't like it.

Comment (ikalnitsky) [inline tests/ext/generators/test_blog.py]: >  I personally didn't like this format at a first glance, I think it decreases readability

That's exactly what I was felt. Now, it's a bit easy, still I'm not sure. I just was tried to reduce a conf size. I mean, it's long and it takes some time to scroll it. Ok, let's do not use this approach then.

File: tests/ext/generators/test_blog.py (hunk 18 above)
Comment (PropheticBird) [inline tests/ext/generators/test_blog.py]: I think this has to be deleted, since these variables relate to the `base.html` template and they are taken from the jinja2 `env` variable. I think,  it's better to make a separate test for the template.

File: tests/ext/generators/test_blog.py (hunk 17 above)
Comment (PropheticBird) [inline tests/ext/generators/test_blog.py]: Issue created.

File: tests/ext/generators/test_blog.py (hunk 19 above)
Comment (PropheticBird) [inline tests/ext/generators/test_blog.py]: OK.
